x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/NativeMemoryCalculatorTests.java

```diff
@@ -0,0 +1,124 @@
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.OptionalLong;
+
+import static org.elasticsearch.xpack.ml.MachineLearning.MACHINE_MEMORY_NODE_ATTR;
+import static org.elasticsearch.xpack.ml.MachineLearning.MAX_JVM_SIZE_NODE_ATTR;
+import static org.elasticsearch.xpack.ml.MachineLearning.MAX_MACHINE_MEMORY_PERCENT;
+import static org.elasticsearch.xpack.ml.MachineLearning.USE_AUTO_MACHINE_MEMORY_PERCENT;
+import static org.hamcrest.Matchers.equalTo;
+
+public class NativeMemoryCalculatorTests extends ESTestCase{
+
+    private static final int NUM_TEST_RUNS = 10;
+    public void testAllowedBytesForMLWhenAutoIsFalse() {
+        for (int i = 0; i < NUM_TEST_RUNS; i++) {
+            long nodeSize = randomLongBetween(100, 10_000);
+            int percent = randomIntBetween(5, 200);
+            DiscoveryNode node = newNode(randomBoolean() ? null : randomNonNegativeLong(), nodeSize);
+            Settings settings = newSettings(percent, false);
+            ClusterSettings clusterSettings = newClusterSettings(percent, false);
+
+            long expected = nodeSize * percent / 100;
+
+            assertThat(NativeMemoryCalculator.allowedBytesForMl(node, settings).getAsLong(), equalTo(expected));
+            assertThat(NativeMemoryCalculator.allowedBytesForMl(node, clusterSettings).getAsLong(), equalTo(expected));
+            assertThat(NativeMemoryCalculator.allowedBytesForMl(node, percent, false).getAsLong(), equalTo(expected));
+        }
+    }
+
+    public void testAllowedBytesForMlWhenAutoIsTrue() {
+        for (int i = 0; i < NUM_TEST_RUNS; i++) {
+            long nodeSize = randomLongBetween(ByteSizeValue.ofMb(500).getBytes(), ByteSizeValue.ofGb(64).getBytes());
+            long jvmSize = randomLongBetween(ByteSizeValue.ofMb(250).getBytes(), nodeSize - ByteSizeValue.ofMb(200).getBytes());
+            int percent = randomIntBetween(5, 200);
+            DiscoveryNode node = newNode(jvmSize, nodeSize);
+            Settings settings = newSettings(percent, true);
+            ClusterSettings clusterSettings = newClusterSettings(percent, true);
+
+            int truePercent = Math.min(
+                90,
+                (int)Math.ceil(((nodeSize - jvmSize - ByteSizeValue.ofMb(200).getBytes()) / (double)nodeSize) * 100.0D));
+            long expected = nodeSize * truePercent / 100;
+
+            assertThat(NativeMemoryCalculator.allowedBytesForMl(node, settings).getAsLong(), equalTo(expected));
+            assertThat(NativeMemoryCalculator.allowedBytesForMl(node, clusterSettings).getAsLong(), equalTo(expected));
+            assertThat(NativeMemoryCalculator.allowedBytesForMl(node, percent, true).getAsLong(), equalTo(expected));
+        }
+    }
+
+    public void testAllowedBytesForMlWhenAutoIsTrueButJVMSizeIsUnknown() {
+        long nodeSize = randomLongBetween(100, 10_000);
```

Comment:
These are bytes, not megabytes.  The test would be testing more realistic size ranges if this was multiplied by `1024 * 1024`.